druid-shell/src/mouse.rs

```diff
@@ -18,50 +18,240 @@ use crate::kurbo::Point;
+    /// The button that was pressed down in the case of mouse-down,
+    /// or the button that was released in the case of mouse-up.
     pub button: MouseButton,
 }
 
 /// An indicator of which mouse button was pressed.
 #[derive(PartialEq, Eq, Clone, Copy, Debug)]
+#[repr(u8)]
 pub enum MouseButton {
     /// Left mouse button.
     Left,
-    /// Middle mouse button.
-    Middle,
     /// Right mouse button.
     Right,
+    /// Middle mouse button.
+    Middle,
     /// First X button.
     X1,
     /// Second X button.
     X2,
+    /// Some other unknown button.
+    Other,
 }
 
 impl MouseButton {
-    /// Returns `true` if this is the left mouse button.
-    #[inline(always)]
+    /// Returns `true` if this is [`MouseButton::Left`].
+    ///
+    /// [`MouseButton::Left`]: #variant.Left
+    #[inline]
     pub fn is_left(self) -> bool {
         self == MouseButton::Left
     }
 
-    /// Returns `true` if this is the right mouse button.
-    #[inline(always)]
+    /// Returns `true` if this is [`MouseButton::Right`].
+    ///
+    /// [`MouseButton::Right`]: #variant.Right
+    #[inline]
     pub fn is_right(self) -> bool {
         self == MouseButton::Right
     }
+
+    /// Returns `true` if this is [`MouseButton::Middle`].
+    ///
+    /// [`MouseButton::Middle`]: #variant.Middle
+    #[inline]
+    pub fn is_middle(self) -> bool {
+        self == MouseButton::Middle
+    }
+
+    /// Returns `true` if this is [`MouseButton::X1`].
+    ///
+    /// [`MouseButton::X1`]: #variant.X1
+    #[inline]
+    pub fn is_x1(self) -> bool {
+        self == MouseButton::X1
+    }
+
+    /// Returns `true` if this is [`MouseButton::X2`].
+    ///
+    /// [`MouseButton::X2`]: #variant.X2
+    #[inline]
+    pub fn is_x2(self) -> bool {
+        self == MouseButton::X2
+    }
+
+    /// Returns `true` if this is [`MouseButton::Other`].
+    ///
+    /// [`MouseButton::Other`]: #variant.Other
+    #[inline]
+    pub fn is_other(self) -> bool {
+        self == MouseButton::Other
+    }
+}
+
+/// A set of [`MouseButton`]s.
+///
+/// [`MouseButton`]: enum.MouseButton.html
+#[derive(PartialEq, Eq, Clone, Copy, Default)]
+pub struct MouseButtons(u8);
+
+impl MouseButtons {
+    /// Create a new empty set.
+    #[inline]
+    pub fn new() -> MouseButtons {
+        MouseButtons(0)
+    }
+
+    /// Add the `button` to the set.
+    #[inline]
+    pub fn add(&mut self, button: MouseButton) {
+        self.0 |= 1 << button as u8;
+    }
+
+    /// Remove the `button` from the set.
+    #[inline]
+    pub fn remove(&mut self, button: MouseButton) {
+        self.0 &= !(1 << button as u8);
+    }
+
+    /// Builder-style method for adding the `button` to the set.
+    #[inline]
+    pub fn with(mut self, button: MouseButton) -> MouseButtons {
+        self.0 |= 1 << button as u8;
+        self
+    }
+
+    /// Builder-style method for removing the `button` from the set.
+    #[inline]
+    pub fn without(mut self, button: MouseButton) -> MouseButtons {
```

Comment:
I'm getting some YAGNI vibes here; it isn't clear when or where these would be used? Are they just by us, or do we expect external users to need to synthesize mouse buttons?

Reply:
Well this specific `without` method is used in the GTK platform code already.

More generally these functions can be used by external users. Yes they might synthesize `MouseButtons` from some sort of hotkey configuration file and then do a check during a mouse event against that.